pulsar-client-cpp/lib/ConsumerImpl.cc

```diff
@@ -170,23 +172,25 @@ void ConsumerImpl::connectionOpened(const ClientConnectionPtr& cnx) {
     cnx->registerConsumer(consumerId_, shared_from_this());
 
     Lock lockForMessageId(mutexForMessageId_);
-    Optional<MessageId> firstMessageInQueue = clearReceiveQueue();
-    if (subscriptionMode_ == Commands::SubscriptionModeNonDurable) {
-        // Update startMessageId so that we can discard messages after delivery
-        // restarts
-        startMessageId_ = firstMessageInQueue;
-    }
-    const auto startMessageId = startMessageId_;
+    // Update startMessageId so that we can discard messages after delivery restarts
+    startMessageId_ = clearReceiveQueue();
+    const auto subscribeMessageId = (subscriptionMode_ == Commands::SubscriptionModeNonDurable)
+                                        ? startMessageId_.get()
+                                        : Optional<MessageId>::empty();
     lockForMessageId.unlock();
```

Comment:
```suggestion
    // Update startMessageId so that we can discard messages after delivery restarts
    const auto startMessageId = clearReceiveQueue();
    lockForMessageId.unlock();
    const auto subscribeMessageId = (subscriptionMode_ == Commands::SubscriptionModeNonDurable)
                                        ? startMessageId
                                        : Optional<MessageId>::empty();
    startMessageId_ = startMessageId;
```

The combination of `Synchronized::operator=` and `Synchronized::get` is not atomic. We can use a local variable to save the value to update.